[ShrinkBorrowScope] Adopt BackwardReachability.

nate-chandler opened the pull request:
Replaced ShrinkBorrowScope's own data flow with the general BackwardReachability.

Took this opportunity to refactor and document the utility.

Taken together these changes make ShrinkBorrowScope serve as a template for a future LexicalDestroyHoisting which will operate on owned lexical values (rather than guaranteed as here) and hoist destroy_values (rather than end_borrows as here) but should otherwise be quite similar.

nate-chandler committed:
[ShrinkBorrowScope] Adopt BackwardReachability.

Replaced ShrinkBorrowScope's own data flow with the general
BackwardReachability.

Took this opportunity to refactor and document the utility.

Taken together these changes make ShrinkBorrowScope serve as a template
for a future LexicalDestroyHoisting which will operate on owned lexical
values (rather than guaranteed as here) and hoist destroy_values (rather
than end_borrows as here) but should otherwise be quite similar.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

atrick commented on lib/SILOptimizer/Utils/ShrinkBorrowScope.cpp:
```diff
@@ -24,295 +33,371 @@ using namespace swift;
-                    SmallVectorImpl<CopyValueInst *> &modifiedCopyValueInsts)
-      : introducer(bbi), deleter(deleter),
-        modifiedCopyValueInsts(modifiedCopyValueInsts), madeChange(false) {}
+  bool containsEnd(SILInstruction *end) const {
+    return llvm::find_if(ends, [&](auto pair) -> bool {
+             return pair.second == end;
+           }) != ends.end();
```
```suggestion
  bool containsEnd(SILInstruction *end) const {
    auto it = ends.find(end.parent());
    return it != end.end() && it->second == end
```

atrick commented on lib/SILOptimizer/Utils/ShrinkBorrowScope.cpp:
```diff
@@ -24,295 +33,371 @@ using namespace swift;
+    result.barriers.push_back(instruction);
+    return true;
+  }
+  if (uses.containsEnd(instruction)) {
```
I haven't figure out yet why you need the `containsEnd` method. If reachability starts backward from the scope ending instructions, then it should never need to check for them.

atrick replied:
It was just implemented that way for simplicity and convenience at the time. I suggest changing reachability so it doesn't check the last use for a barrier. The client can always do that explicitly if needed. Then `initLastUse` should have a comment that reachability assumes it is a non-barrier, because that is pretty subtle.

Alternatively, just add the end_borrows to the users set.

`checkReachableBarrier` is the critical path for the data flow. It should only do constant-time checks and access as few data structures as possible.

nate-chandler replied:
> checkReachableBarrier is the critical path for the data flow. It should only do constant-time checks and access as few data structures as possible.

Makes sense!  I'll see about tweaking reachability like that.

nate-chandler replied:
Made that change along with some others to deal with barrier terminators in https://github.com/apple/swift/pull/41358/commits/05e3c5a8e84e87886652f5f9d1c6e0c7276a7ddd .

nate-chandler committed:
[SILOpt] Reachability handles phis.

Previously, Reachability assumed that phis were not barriers.  Here,
handling for barrier phis is added.  To that end, a new delegate
callback `checkReachableBarrier(PhiValue)` is added.  Before marking the
beginning of a block as reached (or any of its predecessors), check
whether each argument that is a phi is a barrier.  If any is, then
reachability is done.

Implemented the new method in SSADestroyHoisting by splitting apart the
classification of an instruction and the work to do in response to
visiting an instruction.  Then, when visiting a PhiValue, just check
whether any of the predecessors terminators are classified as barriers.
That way, seeing that they're classified that way doesn't result in
noting down that those terminators had been reached (which indeed they
will not have been if any of the terminators from which the values are
flowing into the phi are barriers).

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test source compatibility

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
[SIL] Added SILBasicBlock::hasPhi.

nate-chandler commented:
@swift-ci please test

atrick reviewed (approved):
💯
